### PR TITLE
Membership form - tidy up receipt code

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Membership form email receipt tidy up

Before
----------------------------------------
Code messier

After
----------------------------------------
Code less messy

Technical Details
----------------------------------------
Details all in https://github.com/civicrm/civicrm-core/pull/12651#issuecomment-414200649 - basically I am proposing this as an alternate cleanup approach to using a static function for mailing on the Renewal form

Comments
----------------------------------------

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton Test failures are related `Missing argument 1 for CRM_Member_Form_Membership::emailMembershipReceipt()`

**Comment by eileenmcnaughton** [comment]
@mattwire fixed

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Member/Form/Membership.php [hunk 1]
```diff
@@ -946,50 +946,34 @@ private function convertIsOverrideValue() {
...
     // retrieve 'from email id' for acknowledgement
     $receiptFrom = CRM_Utils_Array::value('from_email_address', $formValues);
 
+    // @tod figure out how much of the stuff below is genuinely shared with the batch form & a logical shared place.
     if (!empty($formValues['payment_instrument_id'])) {
       $paymentInstrument = CRM_Contribute_PseudoConstant::paymentInstrument();
       $formValues['paidBy'] = $paymentInstrument[$formValues['payment_instrument_id']];
     }
 
-    // retrieve custom data
```
Comment (eileenmcnaughton) [inline CRM/Member/Form/Membership.php]: moved to getCustomValuesForReceipt which is called from the form function (but not the function shared with the BatchEntry task)

File: CRM/Member/Form/Membership.php [hunk 2]
```diff
@@ -946,50 +946,34 @@ private function convertIsOverrideValue() {
...
    *
    * @return bool
    *   true if mail was sent successfully
    */
-  public static function emailReceipt(&$form, &$formValues, &$membership) {
+  public static function emailReceipt(&$form, &$formValues, &$membership, $customValues = NULL) {
     // retrieve 'from email id' for acknowledgement
     $receiptFrom = CRM_Utils_Array::value('from_email_address', $formValues);
 
+    // @tod figure out how much of the stuff below is genuinely shared with the batch form & a logical shared place.
```
Comment (mattwire) [inline CRM/Member/Form/Membership.php]: @eileenmcnaughton Typo @tod => @todo :-)

File: CRM/Member/Form/Membership.php [hunk 3]
```diff
@@ -1947,4 +1933,53 @@ protected function isUpdateToExistingRecurringMembership() {
...
+   *
+   * @param array $formValues
+   * @param \CRM_Member_BAO_Membership $membership
+   * @return array
+   */
+  protected function emailMembershipReceipt($formValues, $membership) {
+    $customValues = $this->getCustomValuesForReceipt($formValues, $membership);
+
+    $mailSend = self::emailReceipt($this, $formValues, $membership, $customValues);
+    return [$membership, $mailSend];
```
Comment (mattwire) [inline CRM/Member/Form/Membership.php]: Are these return values ever used?

File: CRM/Member/Form/Membership.php [hunk 4]
```diff
@@ -1700,7 +1686,7 @@ public function submit() {
       $formValues['receipt_text_signup'] = $formValues['receipt_text'];
       // send email receipt
       $this->assignBillingName();
-      $mailSend = self::emailReceipt($this, $formValues, $membership);
+      $mailSend = $this->emailMembershipReceipt($formValues, $membership);
```
Comment (mattwire) [inline CRM/Member/Form/Membership.php]: I don't think $mailSend is ever used?  Also, emailMembershipReceipt is returning an array and $mailSend is the second element...

**Review by mattwire — CHANGES_REQUESTED** [review]
@eileenmcnaughton Not sure about return parameters from emailMembershipReceipt()

File: CRM/Member/Form/Membership.php (hunk 4 above)
Comment (mattwire) [reply CRM/Member/Form/Membership.php]: ok

**Comment by mattwire** [comment]
Tested this and reviewed code.  Still works as before but code is cleaner. Merge on pass

**Comment by seamuslee001** [comment]
Test fails unrelated merging as per the tag

**Pull request merged**